[WIP] Operation expand pr

jjsjann123 opened the pull request:
Tracking our progress on this and heading towards the next PR

jjsjann123 committed:
refactor registration

kevinstephano committed:
Add new Unary and Binary Ops as well as compound ops like lerp, addcmul, add_alpha, and sub_alpha.

kevinstephano committed:
merge with failing tests

jjsjann123 committed:
operatoins filled

jjsjann123 committed:
fixing multiplication through WAR, we should re-examine the order of our op declaration / or revisit type promotion

kevinstephano committed:
Update type.h to have fixed placement of Mod in enum.

kevinstephano committed:
Adding CondtionalOp to support where, clamp, threshold ops.

kevinstephano committed:
Switch up names and arguments for ConditionalOp to TernaryOp.

kevinstephano committed:
Add clamp, threshold, and where. Fixed remainder.

kevinstephano committed:
Add test for Ternary Ops.

jjsjann123 committed:
tests working

jjsjann123 committed:
removing some prints

jjsjann123 committed:
quick fix for remainder and added it to python test

kevinstephano committed:
Added RandLike but without testing.

kevinstephano committed:
Updated UnaryOps test to also test RandLike.

jjsjann123 committed:
[CUDA FUSER] fallback execution added to catch codegen error.

kevinstephano committed:
Added Casting support from float2half and half2float but it fails tests.

jjsjann123 committed:
temporarily disable binary operations

jjsjann123 committed:
monkey shape inference in PyTorch Jit IR done

kevinstephano committed:
Fix up cast ops.  There is still a problem with FP16 intrinsics that has to be solved either by defining them all internal to the fuser or somehow including a CUDA path.

kevinstephano committed:
Added the hand instrinsics for FP16.  Commented out the unused Cast Unary Op Type.

jjsjann123 committed:
disabled output size check for fusion, disabled a hanging python test that Christian is working on a fix

csarofeen committed:
Fix circular compute at references. Extra error checking on codegen usage.

csarofeen committed:
Clang.

jjsjann123 committed:
re-enabling multi output tests

kevinstephano committed:
Move code that was written in header to kernel_resource_strings.

kevinstephano committed:
Add Bool DataType support. No operations use it yet.

jjsjann123 committed:
randn working in python

jjsjann123 committed:
repro attempt 2 _ code lowering

jjsjann123 committed:
integrate clamp / threshold; fix weighted add / sub

jjsjann123 committed:
Added python tests for expanded operation support

Disable Python & Cpp tests for codegen bug temporarily

kevinstephano committed:
Hook up Bool DataType to Ops that will use it.

kevinstephano committed:
Fix gelu to match Aten's impl.  Update FusionUnaryOps with functional matching with Aten execution.

jjsjann123 committed:
added comparison operations

jjsjann123 committed:
patch some missing unary operation

jjsjann123 committed:
added in python integration. updated operand 0 dtype from int to bool in compliance with PyTorch API

kevinstephano committed:
Fixed up runTestKernel to handle Random usage. Added a randLike test.

kevinstephano committed:
Update BinaryOp tests with functional testing against pytorch.

kevinstephano committed:
Updated tests with functional test comparison against Aten.  Codes needs to be simplified as it is brute forced for now.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+
+    std::vector<IterDomain*> dom;
+    for (int i = 0; i < 2; i++)
+      dom.push_back(new IterDomain(new Int(0), new Int()));
```
You can use makeDummyTensor like all other tests.

kevinstephano replied:
I am changing this.  Just need to augment makeDummyTensor() to take a type as well.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+    at::acos,
+    at::asin,
+    at::atan,
+    //at::hardatanh,
```
are we waiting on hardatanh/hardtanh? I see you test randlike later separately, please add a comment here mentioning that.

kevinstephano replied:
I couldn't find an atanh function in ATen.  So, I have commented it to reflect that.  tanh is added back.  Hardtanh or hardatanh is not appropriate.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+
+    auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+    at::Tensor input1     = at::rand({1,4}, options);
```
Should we do something closer to a full wave? Test wouldn't take significantly longer if we do so.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+    auto options_out = at::TensorOptions().dtype(at::kBool).device(at::kCUDA, 0);
+
+    at::Tensor input1     = at::rand({1,4}, options);
+    at::Tensor input2     = at::rand({1,4}, options);
```
Again should we do a larger size?

kevinstephano replied:
It is now trivial to change.  I put my suggestion above.

```
  std::for_each(ops.begin(), ops.end(), [](OpTuple &op) {
    test_op(/*blocks*/ 640, /*threads*/ 64, std::get<2>(op),                                                                           
            /*Aten Func   */ [&op](std::array<IValue,1> &vals) { return std::get<0>(op)(vals[0].toTensor()); },
            /*JIT  Func   */ [&op](Val* in1)->Val* { return unaryOp(std::get<1>(op), in1); },
            /*Output      */ std::make_pair(ValType::TensorView, DataType::Float),    
            /*Inputs Tuple*/ std::make_tuple(std::make_pair(ValType::TensorView, DataType::Float)));
  });
```

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+
+void testGPU_FusionBinaryMathOps() {
+  std::vector<at::Tensor(*)(const at::Tensor&, const at::Tensor&)> aten_funcs {
+    //at::add,
```
I guess we're testing add and sub a lot in other cases, but does it matter if we test them here too?

jjsjann123 replied:
add and sub are actually ternary ops, so API wise it'll be tricky to add them here.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+
+    auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+    at::Tensor input1     = at::rand({1,4}, options);
```
again here

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+
+    std::vector<IterDomain*> dom;
+    for (int i = 0; i < 2; i++)
+      dom.push_back(new IterDomain(new Int(0), new Int()));
```
Again use makeDummyTensor

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+
+    std::vector<IterDomain*> dom;
+    for (int i = 0; i < 2; i++)
+      dom.push_back(new IterDomain(new Int(0), new Int()));
```
Again use makeDummyTensor

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+
+    std::vector<IterDomain*> dom;
+    for (int i = 0; i < 2; i++)
+      dom.push_back(new IterDomain(new Int(0), new Int()));
```
again

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+    auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+
+    at::Tensor input1     = at::rand({1,4}, options);
+    at::Tensor input2     = at::rand({1,4}, options);
```
again

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+      dom.push_back(new IterDomain(new Int(0), new Int()));
+
+    TensorView* tv0 = new TensorView(new TensorDomain(dom), DataType::Float);
+    TensorView* tv1 = new TensorView(new TensorDomain(dom), DataType::Float);
```
again

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+
+    at::Tensor input1     = at::rand({1,4}, options);
+    at::Tensor input2     = at::rand({1,4}, options);
+    at::Tensor input3     = at::rand({1,4}, options);
```
again

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+    torch::jit::fuser::cuda::compileKernel(fusion, &prog);
+    torch::jit::fuser::cuda::runTestKernel(&prog, inputs, outputs);
+
+    ref_output = at::addcmul(input1, input2, input3, 1.f);
```
Can you put a factor that's not identity?

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+    Float*       f0 = new Float(0.f);
+    Float*       f1 = new Float(1.f);
+    TensorView* tv0 = new TensorView(new TensorDomain(dom), DataType::Float);
+    TensorView* tv1 = static_cast<TensorView*>(clamp(tv0, f0, f1));
```
Doesn't rand generate values from 0-1, shouldn't we clamp to something within that range? Or why don't we do something like fuse (inp-0.25)*2?

jjsjann123 replied:
yep, we probably should update the range.

IIUC, this is the unit test where we intended to keep thing as simple as possible. Hence we don't want to add other binary ops into the equation. cc'ing @kevinstephano in case I got anything wrong.

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+    Float*       f0 = new Float(0.f);
+    Float*       f1 = new Float(1.f);
+    TensorView* tv0 = new TensorView(new TensorDomain(dom), DataType::Float);
+    TensorView* tv1 = static_cast<TensorView*>(threshold(tv0, f0, f1));
```
Shouldn't we threshold at something like 0.5?

csarofeen commented on test/cpp/jit/test_gpu.cpp:
```diff
@@ -1067,13 +1150,691 @@ void testGPU_FusionLoopUnroll() {
+  at::manual_seed(0);
+  at::Tensor ref_output = at::rand_like(input1);
+
+  TORCH_CHECK(output.equal(ref_output),
```
I don't know why, but matching rand_like seems very cool to me :-)

jjsjann123 replied:
Note to myself: This is the part I'm so scared about. I think it will break once I bump up the size of the tensor.

csarofeen commented on test/test_jit_cuda_fuser.py:
```diff
@@ -136,7 +147,115 @@ def t(x : torch.Tensor, y : torch.Tensor, z : torch.Tensor):
-        self.assertFalse(self._has_cuda_fusion_group(t_jit.graph_for(x, y, z)))
+        self.assertTrue(self._has_cuda_fusion_group(t_jit.graph_for(x, y, z)))
+
+    @unittest.skipIf(True, "temporary disable for buggy codegen")
```
still buggy?

jjsjann123 replied:
computeAt fix/refactor is in the reduction branch. So yeah still buggy.

csarofeen commented on test/test_jit_cuda_fuser.py:
```diff
@@ -136,7 +147,115 @@ def t(x : torch.Tensor, y : torch.Tensor, z : torch.Tensor):
+            o = operation(o, y)
+            return o
+        t_jit = torch.jit.script(t)
+        x = torch.randn(4, 8, 32, 32, dtype=torch.float, device="cuda")
```
Can we pick less-friendly sizes?

jjsjann123 replied:
Sure.

The rational is that less-friendly sizes are a problem to be solved by random tests. Same rational behind tiny tensors used in majority of cpp/python tests shown here. But I get what you mean and I have no objection against it.

csarofeen commented on test/test_jit_cuda_fuser.py:
```diff
@@ -35,6 +35,16 @@ def tearDown(self):
         torch._C._jit_override_can_fuse_on_gpu(self.old_gpu_fuse)
         super(TestCudaFuser, self).tearDown()
 
+    def _run_helper(self, jit_op, op, should_fuse, *args):
```
Both the c++ tests and Python tests, are testing single operations one by one. I get the feeling we should be trying many more actual fusions across ops. Even if we select 3 ops in order of a list to fuse together and increment through the list I think it would be a major improvement. Either in C++ or Python testing.

jjsjann123 replied:
Note to myself. TODO/Revisit.

csarofeen commented on torch/csrc/jit/codegen/cuda/arith.cpp:
```diff
@@ -120,9 +128,97 @@ TORCH_CUDA_API Val* ceilDiv(Val* v1, Val* v2) {
+      s->getValType().value());
+
+  Val* intrm = binaryOp(BinaryOpType::Mul, v2, s);
+  return binaryOp(BinaryOpType::Add, v1, intrm);
```
Looks good 👍

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:
```diff
@@ -98,6 +98,50 @@ struct TORCH_CUDA_API BinaryOp : public Expr {
+ */
+struct TORCH_CUDA_API TernaryOp : public Expr {
+  ~TernaryOp() = default;
+  TernaryOp(TernaryOpType _type, Val* _out, Val* _in1, Val* _in2, Val* _in3);
```
Do we want to use something like Val* in[3] instead?

csarofeen replied:
or better yet a std::array?

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:
```diff
@@ -98,6 +98,50 @@ struct TORCH_CUDA_API BinaryOp : public Expr {
+  Val* out() const noexcept {
+    return out_;
+  }
+  Val* in1() const noexcept {
```
We already have access to input(i). Maybe we don't explicit functions for all inputs individually.

jjsjann123 replied:
Oops, looks like we are not being consistent there. `Expr` inherits from `IRInputOutput` which provides a pair of `deque<Val*>` for `inputs_` and `outputs_`. But that is not used by any of the operations defined in `ir_internal_nodes.h`.

We need to fix this @kevinstephano

jjsjann123 replied:
NVM, I didn't understand they are just helper pointers. let's leave it as is.

I'll fix the api use. Sorry for the false alarm.

kevinstephano replied:
Generally speaking I think we could make a mechanical Operation class that is perhaps templatized  to cover what seems like very repetitive and mechanical code instead of having specific classes based on the number of inputs.  The additional data member pointers do seem a little redundant now that you point them out.

Perhaps the question is whether to clean up now or later...

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:
```diff
@@ -98,6 +98,50 @@ struct TORCH_CUDA_API BinaryOp : public Expr {
+ private:
+  const TernaryOpType ternary_op_type_;
+  Val* const out_;
+  Val* const in1_;
```
again, maybe we could do Val* const in_[3];

csarofeen commented on test/test_jit_cuda_fuser.py:
```diff
@@ -35,6 +35,16 @@ def tearDown(self):
         super(TestCudaFuser, self).tearDown()
 
+    def _run_helper(self, jit_op, op, should_fuse, *args):
+        torch.cuda.manual_seed_all(123)
```
Are we testing half fusions anywhere?

jjsjann123 replied:
half fusion means float16 or partially fused graph?

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_iostream.cpp:
```diff
@@ -205,7 +236,10 @@ void IRPrinter::handle(const UnaryOp* const uop) {
     os << uop->getUnaryOpType() << "(";
-    handle(uop->in());
+    if(uop->getUnaryOpType() == UnaryOpType::RandLike)
+      os << "rnd";
```
Do we expect this to get hit? Is this just printing "(rnd)"?

jjsjann123 replied:
When you print out the fusion this is what you get.
```
T1[ iblockIdx.x{gridDim.x}, ithreadIdx.x{blockDim.x} ]
   = randLike(rnd);
```
Any concern from this one?

jjsjann123 replied:
The example I used is from the RandLike in unary test. so yeah we would hit this.
But I don't know if there's a better print out in this case. I'll let @kevinstephano confirm this one.

kevinstephano replied:
That is the "operand" in the case of feeding it to the `uniform()` function.  Rnd is the random number generator that gets added to the kernel specifically for random kernels.

jjsjann123 replied:
Just to keep an reference to ourselves.

If we want to attempt mapping RNG output with eager/legacy fuser, we likely would need to have multiple generators and access them per unrolling index.
Having said that, I don't know how we would handle the clamp of grid size per maximum occupancy. :/

csarofeen commented on torch/csrc/jit/codegen/cuda/kernel.cpp:
```diff
@@ -267,6 +276,23 @@ void runKernel(
     kernel_args.push(output);
   }
 
+  // TODO: this probably won't work for us.
```
does it or not?

jjsjann123 replied:
Note to myself. Don't know, will test.

csarofeen commented on torch/csrc/jit/codegen/cuda/kernel.cpp:
```diff
@@ -288,13 +314,21 @@ void runKernel(
+
+  // TODO: we can't randomly clap down this until we got striding.
+  // const auto nBlocks = std::min(entry->max_blocks_, ceilDiv(numel, 128));
+  const auto nBlocks = ceilDiv(numel, 128);
```
Aren't we using unrolling now? Shouldn't it be `ceilDiv( ceilDiv( numel, 128 ), 4 )`

jjsjann123 replied:
This sounds right. Will update.

csarofeen commented on torch/csrc/jit/codegen/cuda/kernel.cpp:
```diff
@@ -308,15 +342,32 @@ void runTestKernel(
     kernel_args.push(output);
   }
 
+  // TODO: this probably won't work for us.
```
Does it or not?

csarofeen commented on torch/csrc/jit/codegen/cuda/kernel_arg.h:
```diff
@@ -27,6 +27,14 @@ struct ArgAbstract {
   virtual void* arg() = 0;
 };
 
+struct ULongArg : public ArgAbstract {
```
mmm, tasty extensibility ;-)

csarofeen commented on torch/csrc/jit/codegen/cuda/parser.cpp:
```diff
@@ -178,17 +164,233 @@ class IrParser {
+          [](const Node* const node,
+             std::unordered_map<size_t, CgValue>& value_map,
+             std::unordered_set<CgValue>& value_set) -> void {
+            static std::unordered_map<Symbol, BinaryOpType> op_mapping({
```
I didn't look through, but I assume someone checked this matched with the codegen list?

csarofeen commented on torch/csrc/jit/codegen/cuda/partition.cpp:
```diff
@@ -23,16 +23,20 @@ static c10::optional<c10::Device> getDevice(const Value* const value) {
   for (auto value : node->outputs()) {
     auto output_device = getDevice(value);
-    if (!output_device.has_value() || output_device.value() != device) {
+    //if (!output_device.has_value() || output_device.value() != device) {
```
remove if not needed?

csarofeen commented on torch/csrc/jit/codegen/cuda/partition.cpp:
```diff
@@ -43,10 +47,10 @@ static bool isFusableDevice(const Node* node, const c10::Device device) {
   auto device = getDevice(node);
   if (!device.has_value()) {
-    return false;
+    //return false;
```
remove if not needed

csarofeen replied:
Uncertain why this is return true though :-/

jjsjann123 replied:
We are returning true when device is unknown. This is trying to resolve cases where shape propagation failed in legacy executor, where device information is not available.

I've been complaining to upstream that their recent type promotion now breaks shape propagation in rand_like that it marks output as cpu tensor

csarofeen commented on torch/csrc/jit/codegen/cuda/partition.cpp:
```diff
@@ -84,33 +71,37 @@ bool isFusableCudaFusionGroup(const Node* const node) {
+    // issue 0: currently codegen doesn't support broadcasting, except in the
+    //          form of stride 0.
+    // We WAR by explicitly extend all tensor to be the broadcasted size. This
+    // however requires that we have identical tensor shape for all output
```
Is this now checked in the shape propagation?

jjsjann123 replied:
No this is currently not checked anywhere.
Checking shape is tricky for two reasons: 1. we could ended up being too conservative on not taking nodes that we should be able to fuse; 2. currently if we are not using profiling executor, there's no guarantee that shape information will be present.

Right now we just blindly accepts nodes in fusion group. If we ended up not being able to generate kernel, we'll fallback to normal interpreter and run it without codegen. (This is the same bahavior legacy fuser has and would make our functional parity comparison easier).

csarofeen commented on torch/csrc/jit/codegen/cuda/type.cpp:
```diff
@@ -31,11 +31,15 @@ ValType promote_type(const ValType& t1, const ValType& t2) {
-    return false;
+    return c10::nullopt;
+  if ((DataType::Half == t1) && (DataType::Float == t2))
+	return UnaryOpType::HalfToFloat;;
```
extra ;

csarofeen reviewed (commented):
Overall looks pretty good to me, nicely done. I believe everything is mostly minor comments.

jjsjann123 commented on torch/csrc/jit/codegen/cuda/arith.cpp:
```diff
@@ -120,9 +128,97 @@ TORCH_CUDA_API Val* ceilDiv(Val* v1, Val* v2) {
   return binaryOp(BinaryOpType::And, v1, v2);
 }
 
+TORCH_CUDA_API Val* add_alpha(Val* v1, Val* v2, Val* s) {
```
NITPICK: this API is not used in parser.
IIUC, the problem is that this API hides `intrm`, which will need to be manually binded to thread/block depends on how loops are nested.

(I don't see this as a problem, this is not a super complicated utility function and parser have similar logic that handles this)

jjsjann123 replied:
same thing for sub_alpha

csarofeen replied:
Actually we should be able to use this interface...
https://github.com/csarofeen/pytorch/blob/50e9c95eb8182cf565c404307c2480c4e8b55192/torch/csrc/jit/codegen/cuda/parser.cpp#L103-L124
These for loops are actually not right, there should only be a single for loop, and it should be based on:
`for (auto val : fusion_->vals())`
This loop will grab every intermediate tensor view in the fusion and set bindings correctly. Then you can use this interface as we're using others.

jjsjann123 committed:
use makeDummyTensor

kevinstephano committed:
Cleanup C++ Operation functional tests.

jjsjann123 committed:
use make dummy

kevinstephano committed:
Cleaned up Single Op Testing.  Updated usage of makeDummyTensor with Single Op testing.

kevinstephano committed:
Remove an extra semi-colon from a line.

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into operation_expand_pr

kevinstephano committed:
Raised allclose absolute tolerance of numerical difference from 1e-08 to 1e-07 to get the remainder function to pass and to allow the binary ops to be more comprehensively tested.

jjsjann123 committed:
update unrolling factor to trim extra blocks;
removed unrolling for rand_like;
  -> note that this still does not work for larger tensor, as native aten kernel
     and legacy fuser has occupancy computation and clamps on grid size.

jjsjann123 committed:
update test size

jjsjann123 committed:
removing temporary shape propagation hack

jjsjann123 committed:
clang-format

jjsjann123 committed:
flake8 fix

jjsjann123 committed:
addressing review comments: a. update transformation on intermediate tensor; b. use ternary arith functions

jjsjann123 committed:
clang-format

jjsjann123 committed:
code fixes

jjsjann123 committed:
clang-format

jjsjann123 committed:
Merge remote-tracking branch 'upstream/master' into operation_expand_pr

kevinstephano committed:
Fixes to pass clang-tidy.

jjsjann123 committed:
default value for unroll_factor_, this fixes cpp tests with runTestKernel

jjsjann123 committed:
clang-format

jjsjann123 committed:
fix flake8

jjsjann123 committed:
debugging CI

jjsjann123 committed:
undo debug changes, fix rand_like test for pascal card in CI
